[JSC] Scan `sourceURL` and `sourceMappingURL` with SIMD

sosukesuzuki opened the pull request:
#### fd25f01b7daf4be5c5cbe0b89ea6b2cb3d7cda7a
<pre>
[JSC] Scan `sourceURL` and `sourceMappingURL` with SIMD
<a href="https://bugs.webkit.org/show_bug.cgi?id=296939">https://bugs.webkit.org/show_bug.cgi?id=296939</a>

Reviewed by Yusuke Suzuki.

JSC&apos;s Lexer currently scans sourceMappingURL and sourceURL directives character
by character. This change introduces SIMD optimization for LChar strings to
improve performance on large inline source maps.

When CharacterType is LChar, JSC treats the following 9 characters as directive
terminators:

- 0x09 (tab) through 0x0D (carriage return): \t, \n, \v, \f, \r
- 0x20 (space)
- 0x22 (double quote)
- 0x27 (single quote)
- 0xA0 (non-breaking space)

The SIMD implementation maintains identical behavior by detecting these same
characters. Characters 0x09-0x0D are detected using range checks for efficiency,
while the other characters are checked individually using bitwise OR operations.

                                    TipOfTree                  Patched

source-mapping-url-short          0.1542+-0.0380            0.1537+-0.0506
source-mapping-url-inlined        1.5875+-0.0425     ^      0.7537+-0.0311        ^ definitely 2.1062x faster

* JSTests/microbenchmarks/source-mapping-url-inlined.js: Added.
(eval.shouldBe):
* JSTests/microbenchmarks/source-mapping-url-short.js: Added.
(eval.shouldBe):
* Source/JavaScriptCore/parser/Lexer.cpp:
(JSC::parseCommentDirectiveValueSIMD):
(JSC::Lexer&lt;CharacterType&gt;::parseCommentDirectiveValue):

Canonical link: <a href="https://commits.webkit.org/298427@main">https://commits.webkit.org/298427@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e207829669dfa4aa0c6f9ddc67b6613e4cbc7727

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/115433 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/35125 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/25626 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/121519 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/66007 "Built successfully") 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/117322 "Build is in progress. Recent messages:") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35785 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/43695 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87704 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42384 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/118381 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28540 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/103610 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68096 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27698 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21732 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/65175 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/107634 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97929 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21845 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/124684 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/113969 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/42369 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31739 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96488 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42736 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99798 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96274 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24498 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41503 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19356 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/38272 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/42249 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/47808 "Built successfully") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/139444 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41752 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/139444 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45079 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/43469 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/parser/Lexer.cpp:
```diff
@@ -1909,16 +1909,57 @@ ALWAYS_INLINE void Lexer<T>::parseCommentDirective()
+        auto squotes = SIMD::equal(input, squoteChar);
+        auto nbsps = SIMD::equal(input, nbspChar);
+
+        auto mask = SIMD::bitOr(SIMD::bitOr(controls, spaces), SIMD::bitOr(quotes, SIMD::bitOr(squotes, nbsps)));
```
You can list them via bitOr(c0, c1, c2, c4, ...)

Constellation commented on Source/JavaScriptCore/parser/Lexer.cpp:
```diff
@@ -1909,16 +1909,57 @@ ALWAYS_INLINE void Lexer<T>::parseCommentDirective()
-            mergedCharacterBits |= m_current;
-        shift();
+    if constexpr (std::is_same_v<CharacterType, LChar>) {
+        m_code = parseCommentDirectiveValueSIMD(m_code, m_codeEnd);
```
Why not using stringStart?

sosukesuzuki committed:
[JSC] Scan `sourceURL` and `sourceMappingURL` with SIMD
https://bugs.webkit.org/show_bug.cgi?id=296939

Reviewed by Yusuke Suzuki.

JSC's Lexer currently scans sourceMappingURL and sourceURL directives character
by character. This change introduces SIMD optimization for LChar strings to
improve performance on large inline source maps.

When CharacterType is LChar, JSC treats the following 9 characters as directive
terminators:

- 0x09 (tab) through 0x0D (carriage return): \t, \n, \v, \f, \r
- 0x20 (space)
- 0x22 (double quote)
- 0x27 (single quote)
- 0xA0 (non-breaking space)

The SIMD implementation maintains identical behavior by detecting these same
characters. Characters 0x09-0x0D are detected using range checks for efficiency,
while the other characters are checked individually using bitwise OR operations.

                                    TipOfTree                  Patched

source-mapping-url-short          0.1542+-0.0380            0.1537+-0.0506
source-mapping-url-inlined        1.5875+-0.0425     ^      0.7537+-0.0311        ^ definitely 2.1062x faster

* JSTests/microbenchmarks/source-mapping-url-inlined.js: Added.
(eval.shouldBe):
* JSTests/microbenchmarks/source-mapping-url-short.js: Added.
(eval.shouldBe):
* Source/JavaScriptCore/parser/Lexer.cpp:
(JSC::parseCommentDirectiveValueSIMD):
(JSC::Lexer<CharacterType>::parseCommentDirectiveValue):

Canonical link: https://commits.webkit.org/298427@main
